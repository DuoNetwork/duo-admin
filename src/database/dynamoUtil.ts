import AWS from 'aws-sdk';
import { PutItemInput } from 'aws-sdk/clients/dynamodb';
import moment from 'moment';
import * as CST from '../constants';
import { IPrice, IPriceBar, ITrade } from '../types';
import util from '../util';

export class DynamoUtil {
	private ddb: undefined | AWS.DynamoDB = undefined;
	// private role: string = '';
	private process: string = 'UNKNOWN';
	private live: boolean = false;
	public init(live: boolean, role: string, process: string) {
		this.live = live;
		this.process = process;
		// this.role = role;
		AWS.config.loadFromPath('./src/keys/aws/' + (live ? 'live' : 'dev') + '/' + role + '.json');
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public insertData(params: PutItemInput): Promise<void> {
		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.putItem(params, err => (err ? reject(err) : resolve()))
					: reject('dynamo db connection is not initialized')
		);
	}

	public convertTradeToSchema(trade: ITrade, systime: number) {
		return {
			[CST.DB_TX_ID]: { S: trade.id },
			[CST.DB_TX_PRICE]: { N: trade.price.toString() },
			[CST.DB_TX_AMOUNT]: { N: trade.amount.toString() },
			[CST.DB_TX_TS]: { N: trade.timestamp + '' },
			[CST.DB_TX_SYSTIME]: { N: systime + '' }
		};
	}

	public convertPriceToSchema(price: IPrice) {
		return {
			[CST.DB_HISTORY_PRICE]: { N: price.price + '' },
			[CST.DB_HISTORY_TIMESTAMP]: { N: price.timestamp + '' },
			[CST.DB_HISTORY_VOLUME]: { N: price.volume + '' }
		};
	}

	public convertPriceBarToSchema(priceBar: IPriceBar) {
		return {
			[CST.DB_HR_SRC_DATE]: { S: priceBar.source + '|' + priceBar.date },
			[CST.DB_HR_HOUR]: { N: priceBar.hour + '' },
			[CST.DB_HR_OPEN]: { N: priceBar.open + '' },
			[CST.DB_HR_HIGH]: { N: priceBar.high + '' },
			[CST.DB_HR_LOW]: { N: priceBar.low + '' },
			[CST.DB_HR_CLOSE]: { N: priceBar.close + '' },
			[CST.DB_HR_VOLUME]: { N: priceBar.volume + '' },
			[CST.DB_HR_TS]: { N: priceBar.timestamp + '' }
		};
	}

	public async insertTradeData(trade: ITrade, insertStatus: boolean): Promise<void> {
		const systemTimestamp = util.getNowTimestamp(); // record down the MTS
		const data = this.convertTradeToSchema(trade, systemTimestamp);

		const params = {
			TableName: this.live ? CST.DB_AWS_TRADES_LIVE : CST.DB_AWS_TRADES_DEV,
			Item: {
				[CST.DB_TX_SRC_DATE]: {
					S: trade.source + '|' + moment.utc(trade.timestamp).format('YYYY-MM-DD')
				},
				...data
			}
		};

		await this.insertData(params);
		if (insertStatus) await this.insertStatusData(data);
	}

	public async insertHourlyData(priceBar: IPriceBar): Promise<void> {
		return this.insertData({
			TableName: this.live ? CST.DB_AWS_HOURLY_LIVE : CST.DB_AWS_HOURLY_DEV,
			Item: {
				...this.convertPriceBarToSchema(priceBar)
			}
		});
	}

	public insertHeartbeat(): Promise<void> {
		return this.insertData({
			TableName: this.live ? CST.DB_AWS_STATUS_LIVE : CST.DB_AWS_STATUS_DEV,
			Item: {
				[CST.DB_ST_PROCESS]: {
					S: this.process
				},
				[CST.DB_ST_TS]: { N: util.getNowTimestamp() + '' }
			}
		});
	}

	public insertStatusData(data: object): Promise<void> {
		return this.insertData({
			TableName: this.live ? CST.DB_AWS_STATUS_LIVE : CST.DB_AWS_STATUS_DEV,
			Item: {
				[CST.DB_ST_PROCESS]: {
					S: this.process
				},
				...data
			}
		});
	}

	public readTradeData(source: string, date: string): Promise<any> {
		const params = {
			TableName: this.live ? CST.DB_AWS_TRADES_LIVE : CST.DB_AWS_TRADES_DEV,
			KeyConditionExpression: CST.DB_TX_SRC_DATE + ' = :' + CST.DB_TX_SRC_DATE,
			ExpressionAttributeValues: {
				[':' + CST.DB_TX_SRC_DATE]: { S: source + '|' + date }
			}
		};

		console.log(params);

		return new Promise(
			(resolve, reject) =>
				this.ddb
					? this.ddb.query(params, (err, data) => (err ? reject(err) : resolve(data)))
					: reject('dynamo db connection is not initialized')
		);
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
