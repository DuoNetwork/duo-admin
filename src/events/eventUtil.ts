import Web3 from 'web3';
import * as CST from '../constants';
import contractUtil from '../contractUtil';
// const provider = 'https://mainnet.infura.io/Ky03pelFIxoZdAUsr82w';
// const provider = 'ws://kovan.infura.io/WSDscoNUvMiL1M7TvMNP';
const provider = 'ws://localhost:8546';
// const provider = 'wss://socket.etherscan.io/wshandler';
const web3 = new Web3(new Web3.providers.WebsocketProvider(provider));
const CustodianABI = require('../static/Custodian.json'); // Custodian Contract ABI
const custodianContract = new web3.eth.Contract(CustodianABI['abi'], CST.CUSTODIAN_ADDR);

let gasPrice: number = 5e9;
const gasLimitPreReset = 120000;
const gasLimitReset = 7880000;

export class EventUtil {
	subscribeToAcceptPrice() {
		console.log('starting listening acceptPrice event');
		custodianContract.events.AcceptPrice(
			{
				fromBlock: CST.INCEPTION_BLK
			},
			function(error, event) {
				if (error) {
					console.log(error);
				} else {
					console.log(event);
				}
			}
		);
	}

	async trigger(abi: object, input: any[], gasPrice: number, gasLimit: number) {
		const nonce = await web3.eth.getTransactionCount(CST.PF_ADDR);
		const command = contractUtil.generateTxString(abi, input);
		// sending out transaction
		web3.eth
			.sendSignedTransaction(
				'0x' +
					contractUtil.signTx(
						contractUtil.createTxCommand(
							nonce,
							gasPrice,
							gasLimit,
							CST.CUSTODIAN_ADDR,
							0,
							command
						),
						CST.PF_ADDR_PK
					)
			)
			.on('receipt', console.log);
	}

	async triggerReset() {
		const abi = {
			name: 'startReset',
			type: 'function',
			inputs: []
		};
		const input = [];
		const web3GasPrice = await contractUtil.getGasPrice();
		gasPrice = web3GasPrice || gasPrice;
		console.log('gasPrice price ' + gasPrice + ' gasLimit is ' + gasLimitReset);
		await this.trigger(abi, input, gasPrice, gasLimitReset);
	}

	async triggerPreReset() {
		const abi = {
			name: 'startPreReset',
			type: 'function',
			inputs: []
		};
		const input = [];
		const web3GasPrice = await contractUtil.getGasPrice();
		gasPrice = web3GasPrice || gasPrice;
		console.log('gasPrice price ' + gasPrice + ' gasLimit is ' + gasLimitPreReset);
		await this.trigger(abi, input, gasPrice, gasLimitPreReset);  // 120000 for lastOne; 30000 for else
	}

	async subscribePreReset() {
		const state = await contractUtil.read('state');
		console.log(state);
		if (state === CST.STATE_PRERESET) await this.triggerPreReset();
		console.log('starting listening preReset event');
		custodianContract.events.StartPreReset(
			{
				fromBlock: CST.INCEPTION_BLK
			},
			async (error, event) => {
				if (error) {
					console.log(error);
				} else {
					console.log(event);
					await this.triggerPreReset();
				}
			}
		);
	}

	async subscribeReset() {
		console.log('starting listening reset event');
		const state = await contractUtil.read('state');
		if (
			state === CST.STATE_UP_RESET ||
			state === CST.STATE_DOWN_RESET ||
			state === CST.STATE_PERIOD_RESET
		) {
			console.log('start triggering reset');
			await this.triggerReset();
		}
		custodianContract.events.StartReset(
			{
				fromBlock: CST.INCEPTION_BLK
			},
			async (error, event) => {
				if (error) {
					console.log(error);
				} else {
					console.log(event);
					await this.triggerReset();
				}
			}
		);
	}
}

const eventUtil = new EventUtil();
export default eventUtil;
