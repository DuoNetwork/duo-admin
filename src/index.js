import pf from "./priceSender/priceFeed";
import listenToAcceptPrice from "./listenToAcceptPrice";
import createAccount from "./accounts/createAccounts";
import bitfinexTradeFeedUtil from "./dataFetcher/bitfinex/bitfinexUtil";
import geminiTradeFeedUtil from "./dataFetcher/gemini/geminiUtil";
import krankenTradeFeedUtil from "./dataFetcher/kraken/krakenUtil";
import coinbaseGDAXTradeFeedUtil from "./dataFetcher/gdax/gdaxUtil";

let tool = process.argv[2];

switch (tool) {
	case "pf":
		console.log("starting commitPrice process");
		pf.startFeeding();
		break;
	case "acceptPrice":
		console.log("starting listening to acceptPrice event");
		listenToAcceptPrice.startListening();
		break;
	case "createAccount":
		console.log("starting create accounts");
		createAccount.createAccount(process.argv[3]);
		break;
	case "bitfinex":
		console.log("starting fetchTrade of bitfinex");
		bitfinexTradeFeedUtil.fetchETHTradesByOwnWebSocket();
		break;
	case "gemini":
		console.log("starting fetchTrade of gemini");
		geminiTradeFeedUtil.fetchETHTradesByOwnWebSocket();
		break;
	case "kraken":
		console.log("starting fetchTrade of kraken");
		krankenTradeFeedUtil.startFetching();
		break;

	case "gdax":
		console.log("starting fetchTrade of gdax");
		coinbaseGDAXTradeFeedUtil.startFetching();
		break;
}
