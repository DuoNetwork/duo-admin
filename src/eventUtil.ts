import ContractUtil from '../../duo-contract-util/src/contractUtil';
import * as CST from './constants';
import dynamoUtil from './database/dynamoUtil';
import { IEvent, IOption } from './types';
import util from './util';

class EventUtil {
	public async subscribe(
		address: string,
		privateKey: string,
		contractUtil: ContractUtil,
		option: IOption
	) {
		util.log('subscribing to ' + option.event);

		if (option.source)
			if ([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event))
				setInterval(async () => {
					const sysState = await contractUtil.getCustodianStates();
					const state = sysState.state;
					util.log('current state is ' + state);

					if (option.event === CST.EVENT_START_PRE_RESET && state === CST.CTD_PRERESET)
						await contractUtil.triggerPreReset(address, privateKey);
					else if (
						option.event === CST.EVENT_START_RESET &&
						[CST.CTD_UP_RESET, CST.CTD_DOWN_RESET, CST.CTD_PERIOD_RESET].includes(
							state
						)
					)
						await contractUtil.triggerReset(address, privateKey);

					dynamoUtil.insertHeartbeat();
				}, 15000);
			else {
				let startBlk = option.force
					? contractUtil.inceptionBlk
					: Math.max(await dynamoUtil.readLastBlock(), contractUtil.inceptionBlk);
				util.log('starting blk number: ' + startBlk);
				let isProcessing = false;
				const fetch = async () => {
					if (isProcessing) return;

					isProcessing = true;
					const currentBlk = await contractUtil.web3.eth.getBlockNumber();
					while (startBlk <= currentBlk) {
						// const block = await contractUtil.web3.eth.getBlock(startBlk);
						// let timestamp = block.timestamp * 1000;
						const allEvents: IEvent[] = [];
						const end = Math.min(startBlk + CST.EVENT_FETCH_BLOCK_INTERVAL, currentBlk);
						const promiseList = CST.EVENTS.map(event =>
							contractUtil.pullEvents(contractUtil.custodian, startBlk, end, event)
						);

						const results = await Promise.all(promiseList);
						for (const result of results)
							for (const el of result) {
								const block = await contractUtil.web3.eth.getBlock(el.blockNumber);
								allEvents.push(contractUtil.parseEvent(el, block.timestamp * 1000));
							}

						util.log(
							'total ' +
								allEvents.length +
								' events from block ' +
								startBlk +
								' to ' +
								end
						);
						if (allEvents.length > 0) await dynamoUtil.insertEventsData(allEvents);
						await dynamoUtil.insertHeartbeat({
							[CST.DB_ST_BLOCK]: { N: end + '' }
						});
						startBlk = end + 1;
					}
					isProcessing = false;
				};
				fetch();
				setInterval(() => fetch(), CST.EVENT_FETCH_TIME_INVERVAL);
			}
		else {
			util.log('starting listening ' + option.event);
			let tg: (r: any) => Promise<any> = () => Promise.resolve();

			if ([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)) {
				const sysState = await contractUtil.getCustodianStates();
				const state = sysState.state;
				util.log('current state is ' + state);

				if (option.event === CST.EVENT_START_PRE_RESET) {
					tg = () => contractUtil.triggerPreReset(address, privateKey);
					if (state === CST.CTD_PRERESET)
						await contractUtil.triggerPreReset(address, privateKey);
				} else if (option.event === CST.EVENT_START_RESET) {
					tg = () => contractUtil.triggerReset(address, privateKey);
					if (
						[CST.CTD_UP_RESET, CST.CTD_DOWN_RESET, CST.CTD_PERIOD_RESET].includes(
							state
						)
					)
						await contractUtil.triggerReset(address, privateKey);
				}
			} else return Promise.resolve();

			contractUtil.custodian.events[option.event]({}, async (error, evt) => {
				if (error) util.log(error);
				else {
					console.log(evt);
					await tg(evt);
				}
			});

			setInterval(() => dynamoUtil.insertHeartbeat(), 30000);
		}
	}
}

const eventUtil = new EventUtil();
export default eventUtil;
