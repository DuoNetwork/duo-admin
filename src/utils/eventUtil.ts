import BaseWrapper from '../../../duo-contract-wrapper/src/BaseWrapper';
import BeethovenWapper from '../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from '../common/constants';
import { IEvent, IOption } from '../common/types';
import dynamoUtil from './dynamoUtil';
import util from './util';

class EventUtil {
	public async trigger(
		address: string,
		privateKey: string,
		beethovenWappers: BeethovenWapper[],
		option: IOption
	) {
		util.logInfo('subscribing to ' + option.event);
		if (![CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)) {
			util.logError('invalid event, exit');
			return Promise.resolve();
		}

		if (option.source)
			setInterval(async () => {
				const promiseList = beethovenWappers.map(async bw => {
					const sysState = await bw.getStates();
					const state = sysState.state;
					util.logDebug('current state is ' + state + ' for ' + bw.address);

					if (option.event === CST.EVENT_START_PRE_RESET && state === CST.CTD_PRERESET)
						await bw.triggerPreReset(address, privateKey);
					else if (option.event === CST.EVENT_START_RESET && state === CST.CTD_RESET)
						await bw.triggerReset(address, privateKey);

					dynamoUtil.insertHeartbeat();
				});
				await Promise.all(promiseList);
			}, 15000);
		else {
			for (const bw of beethovenWappers) {
				util.logInfo('starting listening ' + option.event + ' for ' + bw.address);
				let tg: () => Promise<any> = () => Promise.resolve();
				const sysState = await bw.getStates();
				const state = sysState.state;
				util.logInfo('current state is ' + state + ' for ' + bw.address);

				if (option.event === CST.EVENT_START_PRE_RESET) {
					tg = () => bw.triggerPreReset(address, privateKey);
					if (state === CST.CTD_PRERESET) await tg();
				} else if (option.event === CST.EVENT_START_RESET) {
					tg = () => bw.triggerReset(address, privateKey);
					if (state === CST.CTD_RESET) await tg();
				}

				bw.contract.events[option.event]({}, async (error, evt) => {
					if (error) util.logInfo(error);
					else {
						util.logInfo(evt);
						await tg();
					}
				});
			}

			setInterval(() => dynamoUtil.insertHeartbeat(), 30000);
		}
	}

	public async fetch(baseWrappers: BaseWrapper[], force: boolean) {
		util.logInfo('fetching events');

		const web3Wrapper = baseWrappers[0].web3Wrapper;

		let startBlk = force
			? web3Wrapper.inceptionBlockNumber
			: Math.max(await dynamoUtil.readLastBlock(), web3Wrapper.inceptionBlockNumber);
		util.logInfo('starting blk number: ' + startBlk);
		let isProcessing = false;
		const fetch = async () => {
			if (isProcessing) return;

			isProcessing = true;
			const blockTimestampMap: { [blk: number]: number } = {};
			const currentBlk = await web3Wrapper.getCurrentBlockNumber();
			while (startBlk <= currentBlk) {
				const allEvents: IEvent[] = [];
				const end = Math.min(startBlk + CST.EVENT_FETCH_BLOCK_INTERVAL, currentBlk);
				const promiseList: Array<Promise<any>> = [];
				baseWrappers.forEach(bw =>
					bw.events.forEach(event =>
						promiseList.push(Web3Wrapper.pullEvents(bw.contract, startBlk, end, event))
					)
				);

				const results = await Promise.all(promiseList);
				for (const result of results)
					for (const el of result) {
						const blockTimestamp =
							blockTimestampMap[el.blockNumber] ||
							(await web3Wrapper.getBlockTimestamp(el.blockNumber));
						blockTimestampMap[el.blockNumber] = blockTimestamp;
						allEvents.push(Web3Wrapper.parseEvent(el, blockTimestamp));
					}

				util.logInfo(
					'total ' + allEvents.length + ' events from block ' + startBlk + ' to ' + end
				);
				if (allEvents.length > 0) await dynamoUtil.insertEventsData(allEvents);
				await dynamoUtil.insertHeartbeat({
					[CST.DB_ST_BLOCK]: { N: end + '' }
				});
				startBlk = end + 1;
			}
			isProcessing = false;
		};
		fetch();
		setInterval(() => fetch(), CST.EVENT_FETCH_TIME_INVERVAL);
	}
}

const eventUtil = new EventUtil();
export default eventUtil;
