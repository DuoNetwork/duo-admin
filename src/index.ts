import bitfinexUtil from './apis/bitfinexUtil';
import gdaxUtil from './apis/gdaxUtil';
import geminiUtil from './apis/geminiUtil';
import krakenUtil from './apis/krakenUtil';
import ContractUtil from './contractUtil';
import dynamoUtil from './database/dynamoUtil';
import sqlUtil from './database/sqlUtil';
import dbUtil from './dbUtil';
import eventUtil from './eventUtil';
import util from './util';

const tool = process.argv[2];
util.log('tool ' + tool);
const option = util.parseOptions(process.argv);
util.log('using ' + (option.live ? 'live' : 'dev') + ' env and ' + (option.source || 'local node'));
const contractUtil = new ContractUtil(option);
dbUtil.init(option.dynamo);
dynamoUtil.init(
	option.live,
	util.getDynamoRole(tool, option.dynamo),
	util.getStatusProcess(tool, option)
);
if (['bitfinex', 'gemini', 'kraken', 'gdax', 'commit'].includes(tool) && !option.dynamo) {
	const mysqlAuthFile = require('./keys/mysql.json');
	sqlUtil.init(mysqlAuthFile.host, mysqlAuthFile.user, mysqlAuthFile.password);
}

switch (tool) {
	case 'bitfinex':
		util.log('starting fetchTrade of bitfinex');
		bitfinexUtil.fetchTrades();
		break;
	case 'gemini':
		util.log('starting fetchTrade of gemini');
		geminiUtil.fetchTrades();
		break;
	case 'kraken':
		util.log('starting fetchTrade of kraken');
		krakenUtil.startFetching();
		break;
	case 'gdax':
		util.log('starting fetchTrade of gdax');
		gdaxUtil.startFetching();
		break;
	case 'subscribe':
		eventUtil.subscribe(contractUtil, option);
		setInterval(dynamoUtil.insertHeartbeat, 30000);
		break;
	case 'commit':
		util.log('starting commit process');
		contractUtil.commitPrice(option);
		break;
	case 'minutely':
		dynamoUtil
			.readTradeData('GDAX', '2018-06-01-05-54')
			.then(data => console.log(JSON.stringify(data)));
		break;
	default:
		util.log('no such tool ' + tool);
		break;
}
