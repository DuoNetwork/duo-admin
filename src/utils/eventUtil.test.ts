import { kovan } from '../../../duo-contract-wrapper/src/contractAddresses';
import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import dynamoUtil from './dynamoUtil';
import eventUtil from './eventUtil';

jest.mock('../../../duo-contract-wrapper/src/DualClassWrapper', () =>
	jest.fn(() => ({
		contract: 'dualClassWrapper'
	}))
);

jest.mock('../../../duo-contract-wrapper/src/Web3Wrapper', () =>
	jest.fn(() => ({
		contractAddresses: kovan
	}))
);

test('trigger, worng event name', async () => {
	dynamoUtil.insertHeartbeat = jest.fn();
	global.setInterval = jest.fn();
	const dualClassWrapper = {
		address: 'contractAddress',
		getStates: jest.fn(() =>
			Promise.resolve({
				state: 'state'
			})
		),
		triggerPreReset: jest.fn(),
		triggerReset: jest.fn()
	} as any;
	await eventUtil.trigger('account', [dualClassWrapper], 'event');
	expect(global.setInterval as jest.Mock).not.toBeCalled();
	expect(dynamoUtil.insertHeartbeat as jest.Mock).not.toBeCalled();
});

test('trigger, StartPreReset', async () => {
	dynamoUtil.insertHeartbeat = jest.fn();
	global.setInterval = jest.fn();
	const dualClassWrappers = [
		{
			address: 'contractAddress1',
			getStates: jest.fn(() =>
				Promise.resolve({
					state: 'PreReset'
				})
			),
			triggerPreReset: jest.fn(),
			triggerReset: jest.fn()
		} as any,
		{
			address: 'contractAddress2',
			getStates: jest.fn(() =>
				Promise.resolve({
					state: 'PreReset'
				})
			),
			triggerPreReset: jest.fn(),
			triggerReset: jest.fn()
		} as any
	];
	await eventUtil.trigger('account', dualClassWrappers, 'StartPreReset');
	expect((global.setInterval as jest.Mock).mock.calls[0][1]).toMatchSnapshot();
	await (global.setInterval as jest.Mock).mock.calls[0][0]();

	for (const dcw of dualClassWrappers) {
		expect(dcw.getStates as jest.Mock).toBeCalledTimes(1);
		expect((dcw.triggerPreReset as jest.Mock).mock.calls).toMatchSnapshot();
		expect(dcw.triggerReset as jest.Mock).not.toBeCalled();
	}
	expect(dynamoUtil.insertHeartbeat as jest.Mock).toBeCalledTimes(dualClassWrappers.length);
});

test('trigger, StartReset', async () => {
	dynamoUtil.insertHeartbeat = jest.fn();
	global.setInterval = jest.fn();
	const dualClassWrappers = [
		{
			address: 'contractAddress1',
			getStates: jest.fn(() =>
				Promise.resolve({
					state: 'Reset'
				})
			),
			triggerPreReset: jest.fn(() => Promise.resolve()),
			triggerReset: jest.fn(() => Promise.resolve())
		} as any,
		{
			address: 'contractAddress2',
			getStates: jest.fn(() =>
				Promise.resolve({
					state: 'Reset'
				})
			),
			triggerPreReset: jest.fn(() => Promise.resolve()),
			triggerReset: jest.fn(() => Promise.resolve())
		} as any
	];
	await eventUtil.trigger('account', dualClassWrappers, 'StartReset');
	expect((global.setInterval as jest.Mock).mock.calls[0][1]).toMatchSnapshot();
	await (global.setInterval as jest.Mock).mock.calls[0][0]();

	for (const dcw of dualClassWrappers) {
		expect(dcw.getStates as jest.Mock).toBeCalledTimes(1);
		expect((dcw.triggerReset as jest.Mock).mock.calls).toMatchSnapshot();
		expect(dcw.triggerPreReset as jest.Mock).not.toBeCalled();
	}
	expect(dynamoUtil.insertHeartbeat as jest.Mock).toBeCalledTimes(dualClassWrappers.length);
});

test('trigger, StartReset, wrong state', async () => {
	dynamoUtil.insertHeartbeat = jest.fn();
	global.setInterval = jest.fn();
	const dualClassWrappers = [
		{
			address: 'contractAddress1',
			getStates: jest.fn(() =>
				Promise.resolve({
					state: 'state'
				})
			),
			triggerPreReset: jest.fn(() => Promise.resolve()),
			triggerReset: jest.fn(() => Promise.resolve())
		} as any,
		{
			address: 'contractAddress2',
			getStates: jest.fn(() =>
				Promise.resolve({
					state: 'state'
				})
			),
			triggerPreReset: jest.fn(() => Promise.resolve()),
			triggerReset: jest.fn(() => Promise.resolve())
		} as any
	];
	await eventUtil.trigger('account', dualClassWrappers, 'StartReset');
	expect((global.setInterval as jest.Mock).mock.calls[0][1]).toMatchSnapshot();
	await (global.setInterval as jest.Mock).mock.calls[0][0]();

	for (const dcw of dualClassWrappers) {
		expect(dcw.getStates as jest.Mock).toBeCalledTimes(1);
		expect(dcw.triggerReset as jest.Mock).not.toBeCalled();
		expect(dcw.triggerPreReset as jest.Mock).not.toBeCalled();
	}
	expect(dynamoUtil.insertHeartbeat as jest.Mock).toBeCalledTimes(dualClassWrappers.length);
});

const contractEvents: { [contract: string]: { [event: string]: object } } = {
	contract1: {
		event1: {
			event: 'event1',
			address: 'contract1',
			returnValues: {
				address: 'address1',
				event: 'event1',
				id: 'id1',
				blockHash: 'blockHash1',
				blockNumber: 1001,
				transactionHash: 'txHash1',
				type: 'type1'
			},
			logIndex: 0,
			transactionIndex: 0,
			transactionHash: 'txHash1',
			blockHash: 'blockHash1',
			blockNumber: 1001
		},
		event2: {
			event: 'event2',
			address: 'contract1',
			returnValues: {
				address: 'address2',
				event: 'event2',
				id: 'id2',
				blockHash: 'blockHash2',
				blockNumber: 1002,
				transactionHash: 'txHash2',
				type: 'type2'
			},
			logIndex: 0,
			transactionIndex: 0,
			transactionHash: 'txHash2',
			blockHash: 'blockHash2',
			blockNumber: 1002
		}
	},
	contract2: {
		event1: {
			event: 'event1',
			address: 'contract2',
			returnValues: {
				address: 'address1',
				event: 'event1',
				id: 'id3',
				blockHash: 'blockHash3',
				blockNumber: 1003,
				transactionHash: 'txHash3',
				type: 'type3'
			},
			logIndex: 0,
			transactionIndex: 0,
			transactionHash: 'txHash3',
			blockHash: 'blockHash3',
			blockNumber: 1003
		},
		event2: {
			event: 'event1',
			address: 'contract2',
			returnValues: {
				address: 'address1',
				event: 'event1',
				id: 'id4',
				blockHash: 'blockHash4',
				blockNumber: 1004,
				transactionHash: 'txHash4',
				type: 'type4'
			},
			logIndex: 0,
			transactionIndex: 0,
			transactionHash: 'txHash4',
			blockHash: 'blockHash4',
			blockNumber: 1004
		}
	}
};

const parsedEvents: { [txHash: string]: object } = {
	txHash1: {
		contractAddress: 'contractAddress1',
		type: 'type1',
		id: 'id1',
		blockHash: 'blockHash1',
		blockNumber: 1001,
		transactionHash: 'txHash1',
		logStatus: 'mined',
		parameters: {},
		timestamp: 1234567890
	},
	txHash2: {
		contractAddress: 'contractAddress2',
		type: 'type2',
		id: 'id2',
		blockHash: 'blockHash2',
		blockNumber: 1002,
		transactionHash: 'txHash2',
		logStatus: 'mined',
		parameters: {},
		timestamp: 1234567890
	},
	txHash3: {
		contractAddress: 'contractAddress3',
		type: 'type3',
		id: 'id3',
		blockHash: 'blockHash3',
		blockNumber: 1003,
		transactionHash: 'txHash3',
		logStatus: 'mined',
		parameters: {},
		timestamp: 1234567890
	},
	txHash4: {
		contractAddress: 'contractAddress4',
		type: 'type4',
		id: 'id4',
		blockHash: 'blockHash4',
		blockNumber: 1004,
		transactionHash: 'txHash4',
		logStatus: 'mined',
		parameters: {},
		timestamp: 1234567890
	}
};

test('fetch,', async () => {
	const baseContractWrappers = [
		{
			web3Wrapper: {
				inceptionBlockNumber: 900,
				getCurrentBlockNumber: jest.fn(() => Promise.resolve(1050)),
				getBlockTimestamp: jest.fn((blkNum: number) => Promise.resolve(1234567 * blkNum))
			},
			events: ['event1', 'event2'],
			contract: 'contract1'
		} as any,
		{
			web3Wrapper: {
				inceptionBlockNumber: 900,
				getCurrentBlockNumber: jest.fn(() => Promise.resolve(1050))
			},
			events: ['event1', 'event2'],
			contract: 'contract2'
		} as any
	];

	dynamoUtil.readLastBlock = jest.fn(() => Promise.resolve(999));

	global.setInterval = jest.fn();

	Web3Wrapper.pullEvents = jest.fn((contract, start, end, event) => {
		console.log(start, end);
		return Promise.resolve([contractEvents[contract][event]]);
	});

	Web3Wrapper.parseEvent = jest.fn(el => parsedEvents[el.transactionHash]);

	dynamoUtil.insertEventsData = jest.fn(() => Promise.resolve({}));
	dynamoUtil.insertHeartbeat = jest.fn(() => Promise.resolve());
	await eventUtil.fetch(baseContractWrappers, true);

	expect((Web3Wrapper.parseEvent as jest.Mock).mock.calls).toMatchSnapshot();
});
