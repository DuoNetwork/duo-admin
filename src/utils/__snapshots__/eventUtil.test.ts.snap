// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch, 1`] = `Array []`;

exports[`trigger, StartPreReset 1`] = `15000`;

exports[`trigger, StartPreReset 2`] = `
Array [
  Array [
    "account",
  ],
]
`;

exports[`trigger, StartPreReset 3`] = `
Array [
  Array [
    "account",
  ],
]
`;

exports[`trigger, StartReset 1`] = `15000`;

exports[`trigger, StartReset 2`] = `
Array [
  Array [
    "account",
  ],
]
`;

exports[`trigger, StartReset 3`] = `
Array [
  Array [
    "account",
  ],
]
`;

exports[`trigger, StartReset, wrong state 1`] = `15000`;
